orm/serial_bucket.go

```diff
@@ -0,0 +1,358 @@
+package orm
+
+import (
+	"reflect"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+)
+
+// SerialModel is implemented by any entity that can be stored using SerialModelBucket.
+// GetPrimaryKey/SetPrimaryKey are used to store and access the Key.
+// The ID is always set to nil before serializing and storing the Value.
+type SerialModel interface {
+	weave.Persistent
+	Validate() error
+	GetPrimaryKey() []byte
+	SetPrimaryKey([]byte) error
+}
+
+// SerialModelSlicePtr represents a pointer to a slice of SerialModels. Think of it as
+// *[]SerialModel Because of Go type system, using []SerialModel would not work for us.
+// Instead we use a placeholder type and the validation is done during the
+// runtime.
+type SerialModelSlicePtr interface{}
+
+// SerialModelBucket is implemented by buckets that operates on SerialModels rather than
+// Objects.
+type SerialModelBucket interface {
+	// ByID query the database for a single SerialModel instance by ID. Lookup is done
+	// by the primary index key. Result is loaded into given destination
+	// SerialModel.
+	// This method returns ErrNotFound if the entity does not exist in the
+	// database.
+	// If given SerialModel type cannot be used to contain stored entity, ErrType
+	// is returned.
+	ByID(db weave.ReadOnlyKVStore, key []byte, dest SerialModel) error
+
+	// PrefixScan will scan for all SerialModels with a primary key (ID)
+	// that begins with the given prefix.
+	// The function returns a (possibly empty) iterator, which can
+	// load each SerialModel as it arrives.
+	// If reverse is true, iterates in descending order (highest value first),
+	// otherwise, it iterates in ascending order.
+	PrefixScan(db weave.ReadOnlyKVStore, prefix []byte, reverse bool) (SerialModelIterator, error)
+
+	// ByIndex returns all objects that secondary index with given name and
+	// given key. Main index is always unique but secondary indexes can
+	// return more than one value for the same key.
+	// All matching entities are appended to given destination slice. If no
+	// result was found, no error is returned and destination slice is not
+	// modified.
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest SerialModelSlicePtr) error
+
+	// IndexScan does a PrefixScan, but on the named index. This would let us eg. load all counters
+	// in order of their count. Or easily find the lowest or highest count.
+	IndexScan(db weave.ReadOnlyKVStore, indexName string, prefix []byte, reverse bool) (SerialModelIterator, error)
+
+	// Create saves given SerialModel in the database.
+	//If ID field is provided uses it otherwise generates an auto-incremented key.
+	// Before inserting into database, SerialModel is validated using Validate method.
+	Create(db weave.KVStore, m SerialModel) error
```

Comment:
req: There is no update function defined. With `Create` I have the assumption that it would fail when an id exists already but this is not the case. Therefore rename it to `Save`, `Store`,  `Put`, ...

Reply:
I believe there was once separate cases for Create and Update - one required no key, the other with a key. But it was requested to simplify. I do kind of like that separation to make the code clearer when using it, but it does add some overhead.